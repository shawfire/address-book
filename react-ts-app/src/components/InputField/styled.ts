import styled from '@emotion/styled';
import { theme } from '../../theme';

export const StyledFieldWrapper = styled.div<{
  errorMessage: string;
}>(({ errorMessage }) => ({
  display: 'flex',
  flexDirection: 'column',
  '@media (max-width: 800px)': {
    margin: 'auto 16px',
    maxWidth: '420px'
  },
  input: {
    ...theme.input,
    height: '48px',
    borderRadius: '4px',
    paddingLeft: '16px',
    lineHeight: '24px',
    backgroundColor: errorMessage
      ? theme.colors.errorInputBackgroundColor
      : theme.colors.inputBackgroundColor
  },
  label: {
    ...theme.label,
    height: '24px',
    lineHeight: '20px',
    top: '-2px',
    fontWeight: 'bold',
    marginTop: '22px'
  }
}));

export const StyledCard = styled.div<{
  maxWidth: number;
  backgroundColor: string;
}>(props => ({
  backgroundColor: props.backgroundColor,
  width: '100%',
  maxWidth: props.maxWidth,
  boxShadow:
    '0px 1px 1.5px rgba(0, 0, 0, 0.106), 0px 0px 0.5px rgba(0, 0, 0, 0.08)',
  borderRadius: '3px'
}));
