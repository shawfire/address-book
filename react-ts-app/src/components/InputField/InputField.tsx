import * as React from 'react';
import { StyledFieldWrapper } from './styled';
import { ErrorSvg } from '../Common/ErrorSvg';
import { StyledError } from '../Common/styled';

export interface InputPropTypes {
  id: string;
  name: string;
  type: string;
  'aria-describedby'?: string;
  value: string;
  maxLength?: number;
  autoFocus?: boolean;
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  onBlur: () => void;
  onKeyDown: (e: React.KeyboardEvent<HTMLInputElement>) => void;
  onMouseDown: (e: React.MouseEvent<HTMLInputElement, MouseEvent>) => void;
  ref: React.RefObject<HTMLInputElement>;
}

export interface InputFieldPropTypes {
  id?: string;
  inputProps: InputPropTypes;
  label: string;
  errorMessage: string;
}

function InputField({ inputProps, label, errorMessage }: InputFieldPropTypes) {
  return (
    <StyledFieldWrapper errorMessage={errorMessage}>
      <label htmlFor={inputProps.id}>{label}</label>
      <input {...inputProps} />
      {errorMessage && (
        <StyledError id={`${inputProps.id}-error`}>
          <ErrorSvg />
          &nbsp;
          {errorMessage}
        </StyledError>
      )}
    </StyledFieldWrapper>
  );
}

export default InputField;
