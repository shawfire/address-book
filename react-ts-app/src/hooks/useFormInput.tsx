import * as React from 'react';
import { useState, useRef, useEffect } from 'react';
import { validate, ValidatorsArray } from './validators';
import { applyMask, removeMask, getEndInputValuePos } from './hooks-helpers';
import { InputPropTypes } from '../components/InputField/InputField';

export interface UseFormInputProps {
  initialValue?: string;
  name: string;
  maxLength?: number;
  allowedChars?: string;
  mask?: string;
  maskFillCharRegExp?: RegExp;
  maskCharRegExp?: RegExp;
  validators?: ValidatorsArray;
}

export function useFormInput({
  initialValue = '',
  name,
  maxLength,
  allowedChars = '.',
  mask = '',
  maskFillCharRegExp = /[\s]/,
  maskCharRegExp = /_/g,
  validators = [() => '']
}: UseFormInputProps): [
  InputPropTypes,
  {
    error: string;
    setError: React.Dispatch<any>;
    touched: boolean;
    validate: () => string;
  }
] {
  const [value, setValue] = useState(initialValue);
  const [error, setError] = useState();
  const [touched, setTouched] = useState<boolean>(false);
  const ref: React.RefObject<HTMLInputElement> = useRef(null);
  const allowedCharsRegex = new RegExp(`^${allowedChars}*$`);
  const maxMaskedLength = (mask.match(maskCharRegExp) || []).length;

  // Preserve the cursor position when updating with a mask
  const [curPos, setCurPos] = useState();

  // Used to update the cursor position after a backspace outside the value
  const [goLeft, setGoLeft] = useState<boolean>(false);

  const delKeyCode = 8;
  let delPressed = false;
  let goLeftValue = false; // closure that doesn't cause a re-render

  useEffect(() => {
    resetCursorPosition();
  }, [value]);

  // value does not change
  // but the cursor position still needs to be updated after post render
  useEffect(() => {
    resetCursorPosition();
    goLeftValue = false;
  }, [goLeft]);

  // handle cursor changes in masked input field
  function resetCursorPosition() {
    if (!!mask) {
      if (ref && ref.current) {
        const endInputValuePos = getEndInputValuePos(value, mask);

        if (curPos === 0 || curPos <= endInputValuePos) {
          // return if cursor is inside value
          ref.current.setSelectionRange(curPos, curPos);
          return;
        }
        if (curPos > endInputValuePos) {
          ref.current.setSelectionRange(curPos, curPos);
        } else {
          ref.current.setSelectionRange(endInputValuePos, endInputValuePos);
        }
      }
    }
  }

  // handle backspace on maskedValue
  function onKeyDown(e: React.KeyboardEvent<HTMLInputElement>) {
    if (mask === '') {
      return;
    }
    if (e.keyCode === delKeyCode) {
      delPressed = true;
    }
  }

  function onMouseDown(e: React.MouseEvent<HTMLInputElement, MouseEvent>) {}

  function onChange(e: React.ChangeEvent<HTMLInputElement>) {
    const curPos = e.target.selectionStart;
    const maskedValue = e.target.value;

    const c = curPos && curPos > 0 && maskedValue.charAt(curPos - 1);
    let newCurPos = curPos;
    const endInputValuePos = getEndInputValuePos(value, mask);
    if (delPressed && curPos && curPos > 0) {
      if (maskedValue.charAt(curPos - 1) === ' ') {
        newCurPos && newCurPos--;
      }
      if (curPos >= endInputValuePos) {
        goLeftValue = true;
        setGoLeft(!goLeft);
      }
    } else if (!delPressed && newCurPos) {
      if (newCurPos > endInputValuePos + 1) {
        newCurPos = endInputValuePos + 1;
        if (maskFillCharRegExp.test(maskedValue.charAt(newCurPos - 1))) {
          newCurPos++;
        }
      } else if (maskFillCharRegExp.test(maskedValue.charAt(newCurPos))) {
        newCurPos++;
      }
    }
    setCurPos(newCurPos);
    const newValue = removeMask(maskedValue, mask);
    if (
      allowedCharsRegex.test(newValue) &&
      (mask.length === 0 || newValue.length <= maxMaskedLength)
    ) {
      setValue(applyMask(newValue, mask));
      if (touched) {
        setError(validate(newValue, validators));
      }
    }
    delPressed = false;
  }

  const setFocus = (ref: React.RefObject<HTMLInputElement>) =>
    ref && ref.current && ref.current.focus();

  function onBlur() {
    const error = validate(removeMask(value, mask), validators);
    setError(error);
    setTouched(true);
  }

  function callValidate() {
    const errorMessage = validate(removeMask(value, mask), validators);
    return errorMessage;
  }

  return [
    {
      id: name,
      name: name,
      type: 'text',
      'aria-describedby': `${name}-error`,
      maxLength,
      value,
      onChange,
      onBlur,
      onKeyDown,
      onMouseDown,
      ref
    },
    {
      error,
      setError,
      touched,
      validate: callValidate
    }
  ];
}
