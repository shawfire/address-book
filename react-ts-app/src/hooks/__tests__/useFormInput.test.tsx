import React from "react";
import { ReactWrapper, shallow, ShallowWrapper } from "enzyme";
import { useFormInput } from "../useFormInput";
import InputField from "../../components/InputField/InputField";

function TestHook() {
  const [givenName, givenNameCheck] = useFormInput({
    name: "givenName"
  });
  return (
    <InputField
      label="Given name"
      errorMessage={givenNameCheck.error}
      inputProps={{
        ...givenName
        // autoFocus: true
      }}
    />
  );
}

describe("Input Field tests", () => {
  let wrapper: ReactWrapper | ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<TestHook />);
  });

  test("Input Field snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Input Field state test", () => {});
});
