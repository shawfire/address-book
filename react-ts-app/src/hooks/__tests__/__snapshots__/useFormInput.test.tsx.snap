// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Field tests Input Field snapshot 1`] = `ShallowWrapper {}`;
